### PR TITLE
Improve performance of businessCalendar's add()

### PR DESCRIPTION
**Pull request by RemiMattheyDoret**
# Context

Similar to [edgelaboratories/tarantino#674](https://github.com/edgelaboratories/tarantino/pull/674)

This PR improves the performance of the function `add`.

# Content

Modification to `add` to avoid looping and calling itself. 

Benchmark added

# How we know it works

Tests pass. No new test added

# Benchmark results

### Intro

I considered four shift values `17`, `-17`, `7778` and `-7778`, referred to in the function names as `xxx_17`, `xxx_m17`, `xxx_7778`, `xxx_m7778`. The large values are irealistic and are used only for considerations of extremes.

### Results

version in master

```
BenchmarkShiftBusinessDays_17-8         	87126721	        11.93 ns/op	       0 B/op	       0 allocs/op
BenchmarkShiftminusBusinessDays_m17-8   	127907304	        9.362 ns/op	       0 B/op	       0 allocs/op
BenchmarkShiftBusinessDays_7778-8       	99965977	        11.94 ns/op	       0 B/op	       0 allocs/op
BenchmarkShiftBusinessDays_m7778-8      	100000000	        11.71 ns/op	       0 B/op	       0 allocs/op
```

version in this branch

```
BenchmarkShiftBusinessDays_17-8         	196057020	         6.129 ns/op	       0 B/op	       0 allocs/op
BenchmarkShiftminusBusinessDays_m17-8   	196296846	         6.138 ns/op	       0 B/op	       0 allocs/op
BenchmarkShiftBusinessDays_7778-8       	197015299	         6.086 ns/op	       0 B/op	       0 allocs/op
BenchmarkShiftBusinessDays_m7778-8      	195863832	         6.120 ns/op	       0 B/op	       0 allocs/op
```

### Discussion

The new version is slightly slower than tarantino's. I would assume this is because tarantino avoid using a date.Date. Also, the old version of calendar was not as bad as tarantino's old version. As a consequence, the improvement is not as extreme as observed in [edgelaboratories/tarantino#674](https://github.com/edgelaboratories/tarantino/pull/674).

### CODE REVIEW

File: business_calendar_test.go
Comment (antonioiubatti93) [inline business_calendar_test.go:341]: Factorize everything in benchmark cases without helper (they work exactly as test cases):
```go
calendar := newBusinessCalendar()
for _, bc := range []struct {
   name string
   date date.Date
   shift int
}{
   // etc.
}
```

**Review by antonioiubatti93 — COMMENTED** [review]
Thank you for taking these on :heart: Notice that there's already a benchmark in the general calendar: https://github.com/edgelaboratories/calendar/blob/main/calendar_test.go#L116
Btw looks like a nice improvement :muscle:

File: business_calendar.go
Comment (antonioiubatti93) [inline business_calendar.go:133]: With early return you don't need the else part at all (avoid it in general)
```suggestion
	if days > 0 && int(time.Friday-current.Weekday()) < nbDaysLeft {
		return nbDaysLeft + 2
	}
	
	if days < 0 && int(current.Weekday()-time.Monday) < -nbDaysLeft {
		return nbDaysLeft - 2
	}
```

**Review by antonioiubatti93 — COMMENTED** [review]
Feel free to add `varnamelen` among the list of disabled linters

**Review by antonioiubatti93 — COMMENTED** [review]
@borisdavid could you check the outcome on the existing benchmarks, too? https://github.com/edgelaboratories/calendar/pull/24#pullrequestreview-1015031863

**Comment by borisdavid** [comment]
Here are the existing benchmarks :

- `Benchmark_Calendar_Add`

On `master` :
image

On this branch : 
image

- `Benchmark_Calendar_DaysBetween`

On `master` :
image

On this branch : 
image

**Comment by antonioiubatti93** [comment]
@borisdavid thx for [this](https://github.com/edgelaboratories/calendar/pull/24#issuecomment-1168661243) :wink: `Benchmark_calendar_add` sets are quite consequential (similar gain to the ones added in this PR), but the others don't show any significant impact. That's because the `DaysBetween` function still uses a for loop. I guess we could apply the same change there as well - in this case I'd find reasonable to do it here, as it's the same concept

**Comment by borisdavid** [comment]
I reworked the function `daysBetween` by trying out some other way to compute `businessDays` (hopeful not too creative :sweat_smile:). The `for`-loops contained in this function were quite marginals (not leading to a linear complexity) so the improvement will be minor compared to the function `add`. Here is the new benchmark for `daysBetween` : 
image

File: business_calendar_test.go
Comment (antonioiubatti93) [inline business_calendar_test.go:341]: Not sure it's really necessary, but if for some reason the numbers used in here highlight some particular property keep it (maybe documented)

**Review by antonioiubatti93 — DISMISSED** [review]
Just cosmetics - I'll re-approve once ready

File: business_calendar_test.go
Comment (borisdavid) [reply business_calendar_test.go:341]: That is @RemiMattheyDoret choices for the benchmarks, I do not really know which reasons lead him to choose these numbers...

Comment (antonioiubatti93) [reply business_calendar_test.go:341]: Ok, if that's the case we can keep this on hold and let him complete it :wink: or leave an issue for him, too - there's no pressure on this one

**Review by antonioiubatti93 — APPROVED** [review]

**Comment by antonioiubatti93** [comment]
@borisdavid please leave an issue for @RemiMattheyDoret so that he can concile the benchmarks later on :+1: many thx for your contribution!

**Pull request merged**
